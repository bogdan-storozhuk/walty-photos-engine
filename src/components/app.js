import React from 'react';
import { Route, Switch } from 'react-router-dom';
import 'bootstrap/dist/css/bootstrap.min.css';

import { MainPage, PhotoPageContainer } from '.';

const App = () => (
  <Switch>
    <Route path="/" component={MainPage} exact />
    <Route
      path="/photo-page/:id"
      render={({ match }) => {
        const { id } = match.params;
        return <PhotoPageContainer photoId={id} />;
      }}
      exact
    />
    <Route
      path="/:tag1?/:tag2?/:tag3?/:tag4?/:tag5?/:tag6?/:tag7?/:tag8?/:tag9?/:tag10?"
      render={({ match }) => <MainPage tags={match.params} />}
    />
  </Switch>
);

App.propTypes = {};

export default App;
