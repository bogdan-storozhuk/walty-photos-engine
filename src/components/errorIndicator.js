import React from 'react';

const ErrorIndicator = () => <div>Error!</div>;

ErrorIndicator.propTypes = {};

export default ErrorIndicator;
