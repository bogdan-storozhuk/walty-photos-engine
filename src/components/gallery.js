import React from 'react';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import { GalleryItem } from './';
import { Spinner } from './';
import { ErrorIndicator } from './';

const Gallery = ({ photos }) => (
  <div className="row mt-3">
    {photos.map((photo) => {
      return <GalleryItem key={photo.id} photo={photo} />;
    })}
  </div>
);

const GalleryContainer = () => {
  const photos = useSelector((state) => state.photoReducer.photos),
    loading = useSelector((state) => state.photoReducer.loading),
    error = useSelector((state) => state.photoReducer.error);
  if (loading) {
    return <Spinner />;
  }

  if (error) {
    return <ErrorIndicator />;
  }

  return <Gallery photos={photos} />;
};

Gallery.propTypes = {
  photos: PropTypes.arrayOf(
    PropTypes.shape({
      largeImageURL: PropTypes.string,
      id: PropTypes.number,
    })
  ),
};

export default GalleryContainer;
