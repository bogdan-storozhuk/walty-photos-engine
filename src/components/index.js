import App from './app';
import ErrorBoundry from './errorBoundry';
import ErrorIndicator from './errorIndicator';
import GalleryContainer, { Gallery } from './gallery';
import GalleryItem from './galleryItem';
import LastSearchesPanel from './lastSearchesPanel';
import MainPage from './mainPage';
import SearchPanel from './searchPanel';
import Spinner from './spinner';
import Tag from './tag';
import PhotoPageContainer, { PhotoPage } from './photoPage';

export {
  App,
  MainPage,
  SearchPanel,
  LastSearchesPanel,
  GalleryContainer,
  Gallery,
  GalleryItem,
  Tag,
  ErrorBoundry,
  ErrorIndicator,
  Spinner,
  PhotoPageContainer,
  PhotoPage,
};
