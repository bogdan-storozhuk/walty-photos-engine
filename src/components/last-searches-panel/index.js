import LastSearchesPanel from './last-searches-panel';
export default LastSearchesPanel;