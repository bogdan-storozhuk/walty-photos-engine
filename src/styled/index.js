import GalleryItemContainer from './galleryItemContainer';
import GalleryItemImage from './galleryItemImage';
import TagSpan from './tagSpan';
import StyledLink from './styledLink';

export { GalleryItemContainer, GalleryItemImage, TagSpan, StyledLink };
