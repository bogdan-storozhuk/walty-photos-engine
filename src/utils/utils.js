import checkPropTypes from 'check-prop-types';

const deleteUndefinedValuesFromObject = (obj) => {
  for (var propName in obj) {
    if (obj[propName] === null || obj[propName] === undefined) {
      delete obj[propName];
    }
  }
};

const transformObjectIntoArray = (obj) => {
  return Object.values(obj).map((item) => ({
    name: item,
    id: Date.now() + Math.floor(Math.random() * 1000),
  }));
};

const generateRandomId = () => Date.now() + Math.floor(Math.random() * 1000);

const checkProps = (component, expectedProps, propTypes) => {
  const propsErr = checkPropTypes(
    propTypes,
    expectedProps,
    'props',
    component.name
  );
  return propsErr;
};

export {
  deleteUndefinedValuesFromObject,
  transformObjectIntoArray,
  generateRandomId,
  checkProps,
};
