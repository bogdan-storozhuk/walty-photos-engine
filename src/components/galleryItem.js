import React from 'react';
import PropTypes from 'prop-types';

import { GalleryItemContainer, GalleryItemImage } from '../styled/';

const GalleryItem = ({ photo }) => (
  <GalleryItemContainer className="col-lg-3 col-md-4 col-6 mb-4">
    <GalleryItemImage
      src={photo.largeImageURL}
      alt={photo.id}
    ></GalleryItemImage>
  </GalleryItemContainer>
);

GalleryItem.propTypes = {
  photo: PropTypes.shape({
    largeImageURL: PropTypes.string,
    id: PropTypes.number,
  }),
};

export default GalleryItem;
