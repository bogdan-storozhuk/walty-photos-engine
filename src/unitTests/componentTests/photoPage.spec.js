import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure, mount } from 'enzyme';

import { PhotoPageContainer, PhotoPage } from '../../components';
import {
  photoPagePropTypes,
  photoPageContainerPropTypes,
} from '../../utils/propTypes';
import { checkProps } from '../../utils/utils';
import { StyledImageContainer, StyledImage } from '../../styled/';

configure({ adapter: new Adapter() });

let mockStore = configureMockStore([]);

describe('PhotoPage Component', () => {
  const expectedProps = {
    singlePhoto: {
      id: 1,
      downloads: 255,
      favorites: 333,
      likes: 444,
      views: 5000,
      tags: 'some tag data',
      largeImageURL: 'img url',
    },
  };
  describe('Checking PropTypes', () => {
    it('Should not throw a warning', () => {
      const propsError = checkProps(
        PhotoPage,
        expectedProps,
        photoPagePropTypes
      );
      expect(propsError).toBeUndefined();
    });
  });

  describe('PhotoPage Component initial', () => {
    const photoPage = shallow(<PhotoPage {...expectedProps} />);

    it('Should render initial', () => {
      expect(photoPage.find('Container')).toHaveLength(1);
    });

    it('Should render StyledImageContainer', () => {
      expect(photoPage.find(StyledImageContainer)).toHaveLength(1);
    });

    it('Should render StyledImage', () => {
      expect(photoPage.find(StyledImage)).toHaveLength(1);
    });

    it('Should render singlePhoto.downloads in <td>', () => {
      expect(photoPage.find('tbody').childAt(0).childAt(0).text()).toEqual(
        String(expectedProps.singlePhoto.downloads)
      );
    });

    it('Should render singlePhoto.favorites in <td>', () => {
      expect(photoPage.find('tbody').childAt(0).childAt(1).text()).toEqual(
        String(expectedProps.singlePhoto.favorites)
      );
    });

    it('Should render singlePhoto.likes in <td>', () => {
      expect(photoPage.find('tbody').childAt(0).childAt(2).text()).toEqual(
        String(expectedProps.singlePhoto.likes)
      );
    });

    it('Should render singlePhoto.views in <td>', () => {
      expect(photoPage.find('tbody').childAt(0).childAt(3).text()).toEqual(
        String(expectedProps.singlePhoto.views)
      );
    });

    it('Should render singlePhoto.tags in <td>', () => {
      expect(photoPage.find('tbody').childAt(0).childAt(4).text()).toEqual(
        expectedProps.singlePhoto.tags
      );
    });

    it('Should render properly', () => {
      expect(photoPage).toMatchSnapshot();
    });
  });
});

describe('PhotoPageContainer Component', () => {
  const initialState = {
    photoReducer: {
      searchedPhoto: null,
      photos: [],
      loading: false,
      error: null,
    },
  };
  const expectedProps = {
    photoId: '1',
  };

  describe('Checking PropTypes', () => {
    it('Should not throw a warning', () => {
      const store = mockStore(initialState);
      const wrapper = mount(
        <Provider store={store}>
          <PhotoPageContainer {...expectedProps} />
        </Provider>
      );
      const propsError = checkProps(
        wrapper,
        expectedProps,
        photoPageContainerPropTypes
      );
      expect(propsError).toBeUndefined();
    });
  });

  describe('PhotoPageContainer Loading', () => {
    const nextState = {
      ...initialState,
      photoReducer: {
        ...initialState.photoReducer,
        loading: true,
      },
    };
    const store = mockStore(nextState);
    const wrapper = mount(
      <Provider store={store}>
        <PhotoPageContainer {...expectedProps} />
      </Provider>
    );

    it('Should render Spinner', () => {
      expect(wrapper.find('Spinner')).toHaveLength(1);
    });

    it('Should NOT render <ErrorIndicator />', () => {
      expect(wrapper.find('ErrorIndicator')).toHaveLength(0);
    });

    it('Should NOT render <PhotoPage/>', () => {
      expect(wrapper.find('PhotoPage')).toHaveLength(0);
    });

    it('Should render properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('PhotoPageContainer with errorMsg', () => {
    const nextState = {
      ...initialState,
      photoReducer: {
        ...initialState.photoReducer,
        error: 'some error',
      },
    };
    const store = mockStore(nextState);
    const wrapper = mount(
      <Provider store={store}>
        <PhotoPageContainer {...expectedProps} />
      </Provider>
    );

    it('Should renders ErrorIndicator', () => {
      expect(wrapper.find('ErrorIndicator')).toHaveLength(1);
    });

    it('Should NOT render Spinner', () => {
      expect(wrapper.find('Spinner')).toHaveLength(0);
    });

    it('Should NOT render <PhotoPage/>', () => {
      expect(wrapper.find('PhotoPage')).toHaveLength(0);
    });

    it('Should render properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('PhotoPageContainer render PhotoPage', () => {
    const nextState = {
      ...initialState,
      photoReducer: {
        ...initialState.photoReducer,
        searchedPhoto: {
          id: 1,
          downloads: 255,
          favorites: 333,
          likes: 444,
          views: 5000,
          tags: 'some tag data',
          largeImageURL: 'img url',
        },
      },
    };
    const store = mockStore(nextState);
    const wrapper = mount(
      <Provider store={store}>
        <PhotoPageContainer />
      </Provider>
    );

    it('Should NOT render ErrorIndicator', () => {
      expect(wrapper.find('ErrorIndicator')).toHaveLength(0);
    });

    it('Should NOT render Spinner', () => {
      expect(wrapper.find('Spinner')).toHaveLength(0);
    });

    it('Should render <PhotoPage/>', () => {
      expect(wrapper.find('PhotoPage')).toHaveLength(1);
    });

    it('Should render properly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
