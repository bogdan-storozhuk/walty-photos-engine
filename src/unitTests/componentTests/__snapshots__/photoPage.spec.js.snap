// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotoPage Component PhotoPage Component initial Should render properly 1`] = `ShallowWrapper {}`;

exports[`PhotoPageContainer Component PhotoPageContainer Loading Should render properly 1`] = `ReactWrapper {}`;

exports[`PhotoPageContainer Component PhotoPageContainer render PhotoPage Should render properly 1`] = `ReactWrapper {}`;

exports[`PhotoPageContainer Component PhotoPageContainer with errorMsg Should render properly 1`] = `ReactWrapper {}`;
